### DISCUSSION TITLE
Config formatted sheets

### DISCUSSION THREAD

**Discussion by geekscrapy** [saulpw/visidata#1488] [Ideas]
For those that load the same sheet types repeatedly it might be good to allow to pre-set the column types (and other things like aggs?) in `.visidatarc`.

For example, to always set the "timestamp" column in every sheet to date type:
`Basesheet.columns.timestamp.type = date`

To rename a column:
`Basesheet.columns.prevcolname.name = "timestamp"`

To automatically sort once the sheet has loaded:
`Basesheet.columns.timestamp.sort = 'asc'`

Thoughts? Do people use the same source data repeatedly?

**Comment by frosencrantz** [comment]
I have some common queries I perform around the same set of tables or other datasources.  I've wrapped them up into some scripts.  I typically perform the query, then I might use a VisiData script to reshape the types and sorting.  While I use the VisiData scripting, I wish there was a more declarative way, matching column names to types, formatting, sorting, ordering, and visible/hidden.

I had thought maybe vds format might be possible to use, but that requires formatting that data as json, which only some of my data is already.  This is not a big problem, but haven't tried it yet.

I might be getting away from the idea you were suggesting where you could update your .visidatarc file and say based on a column name alone it is safe to assume the format of the column's type and formatting.  I think someone recently (last few months) responded on an older issue or discussion that they had solved a problem like  by creating their own command they ran after loading data.

**Reply by saulpw** [reply]
I've been thinking about a simplified cmdlog that doesn't specify the sheet/col/row for every command, but has separate commands to go to the specified sheet/col/row.  Then you only need the longname and the input:

```
col num
type-int
rename-col number
```

We can make shorter aliases for the commands, or allow the keystrokes themselves to work.  We could use e.g. `,` instead of newline as the command separator, and an option like `-m` on the CLI to specify this cmdlog string, or `-k` for keystrokes.  Then:

    vd foo.csv -m "col num, type-int, rename-col number"

or even:

    vd foo.csv -k $'cnum\n#^number\n'

How far does this get us?

**Reply by geekscrapy** [reply]
Personally, the reason I mentioned putting it in .rc a specific config file is to make it a set and forget, and easily distributable.
Plus, if you have a lot of columns, defining all these on the command line could be clumsy and a bit error prone

**Comment by frosencrantz** [comment]
I wonder if @skeept's suggestion in https://github.com/saulpw/visidata/issues/349  would be helpful here.

Looks like @skeept uses this:  https://github.com/skeept/dotvim/blob/master/test/keep/visidatarc

skeept adds a `guess-types` command with a `guessType` function.

**Comment by geekscrapy** [comment]
#1681 ref

**Reply by frosencrantz** [reply]
@cool-RR shared  autotune code: https://gist.github.com/cool-RR/0a2222468c5673b4dcb6c772e110620c

https://github.com/saulpw/visidata/issues/1700#issuecomment-1407449758 ref

And a note from Saul at some point it might be possible to auto-trigger: https://github.com/saulpw/visidata/issues/1681#issuecomment-1407597781

**Comment by saulpw** [comment]
Okay, with the new `Sheet.afterLoad` function, @geekscrapy's original request is very doable.  I even added a new DefaultAttrDict to get nearly the exact syntax requested:

    Sheet.knownCols.timestamp.type = date
    DirSheet.knownCols.directory.width = 0

**Reply by geekscrapy** [reply]
Well, this is amazing 😅
